### ISSUE TITLE
SMC^2 State Inference

### ISSUE THREAD

**Issue by lundalana** [nchopin/particles#25]
Thank you for developing such a helpful library!

I had a quick question on the implementation of the SMC^2 algorithm for state and parameter inference. In running the algorithm for my case study, I generate an SMC2 sampler object on the states and use it to run the SMC algorithm on the parameters, in general just using the standard options as described in the accompanying textbook. 

The reports generated in the SMC run give me summaries on the inferred parameters, but I'm unable to find any information (particles, weights, moments, etc.) on the inferred states. Is there any way that I can recover this information?

**Comment by nchopin** [comment]
You're welcome! 

Let's call smc2 your SMC2 object. If you inspect the object
`smc2.X `
that stores the N particles, you will see that it contains various attributes, including:
`smc2.X.pfs`
a list of N_theta SMC objects,  representing the N_theta local particle filters (run in the x-dimension). In particular: 
`smc2.X.pfs[m].X `
will be a numpy array that represents the N_x particles X_T^n,m for a given m (T= final time).  

You could for instance approximate the marginal expectation of X_T (given the full data) as follows:

```
import numpy as np

marginal_exp_xT = np.average([np.average(pf.X, weights=pf.W, axis=0) for pf in smc2.X.pfs], weigths=smc2.W)
```

In case you want to compute the marginal distribution of previous states, x_t, t<T, things get a bit more tricky. One way to do it is to apply a backward sampling step to each local particle filter. Now that you mention it, I should implement it in the library. But it's possible to do it manually, as follows:
```
import particles

Nth = 100  # number of theta particles
# define here fk_smc2
smc2 = particles.SMC(fk=fk_smc2, N=Nth, smc_options={'store_history':True})
smc2.run()

if dx == 1:  # dx = dimension of state X_t
    states = np.empty((Nth, T))
else:
    states = np.empty((Nth, dx, T))
for n, pf in enumerate(smc2.X.pfs):
    states[n, ...] = pf.history.extract_one_trajectory()

```
In short: 
* the first three lines run SMC2, with an option passed to each local filter, so that its stores the full history (the x-particles and their weights at each time). This is expensive, memory-wise. 
* the second part generates state variables from the posterior. 

Note: I did not really test all this, so there might be typos, bugs, etc. Let me know if it works for you.

**Comment by lundalana** [comment]
Thanks for the advice! 

I was able to get the first code segment to work, and I think I understand pretty well how it's extracting and manipulating the stored data at the last time step to find the marginal expectation of X_T. 

I think this could easily be applied during run-time on the algorithm to give the marginal distribution on X_t given the data up to t. It would add a little computation time in each loop, but would definitely cut down on the memory requirements. For example:

```
for i in range(len(tau)):
    next(smc2 )
    meanStates[i] = np.average(np.array([np.average(pf.X, weights=pf.W, axis=0) for pf in smc2 .X.pfs]), 
                 weights=smc2 .W, axis=0) 
    varStates[i] = np.average(np.array([np.average(np.square(pf.X), weights=pf.W, axis=0) for pf in smc2 .X.pfs]), 
                 weights=smc2 .W, axis=0) - np.square(meanStates[i])
```

I had more trouble with the second approach, extracting the marginal distribution of the previous states. I ended up modifying the code you suggested as shown below. 

```
bwFKMod_smc2 = ssp.SMC2(ssm_cls=BoucWen, data=respAcc, prior=my_prior, init_Nx=10, 
                   ar_to_increase_Nx=0.1, **smc_options={'store_history':True}**)

bwSMC2 = particles.SMC(fk=bwFKMod_smc2, N=50, resampling = 'systematic',
                    ESSrmin = 0.5, moments=True, store_history=True)

bwSMC2.run()

infStates = np.empty((50, 3, len(tau))
for n, pf in enumerate(bwSMC2.X.pfs):
    infStates[n,:,:] = np.array(pf.hist.extract_one_trajectory()).transpose()
```
I was a little confused at the purpose of extracting only a single trajectory from each local particle filter. It seems like this would give little information about the marginal distribution, as it would only be picking points from the local particle filters instead of representing the distribution itself. However, my understanding might be incorrect. Could you clarify?

Thanks again!

**Comment by nchopin** [comment]
Hi,
sorry, I was mixing up two different things, which was confusing. 

1. If I understand correctly, you wish to approximate, at each time t, the marginal distribution of *current* state X_t, given the data up to time t? In particular, compute the mean and variance of these marginal distributions? In that case, your first block is the way to go, don't worry about the second part. You could simplify it a little by using helper function resampling.wmean_and_var in the resampling module, but that's anecdotal. 
Do you think this should be implemented by default? i.e. these quantities should be computed and gathered in object smc2.summaries? 

2. The thing about "extracting trajectories" is for the more general (and difficult) problem of approximating the distribution of theta and X_{0:t} (all the states up to time t) at a given time. This point is explained in Section 3.3 of the SMC2 paper:
https://rss.onlinelibrary.wiley.com/doi/full/10.1111/j.1467-9868.2012.01046.x
But, as you noted, if you would do this at every time t, this is likely to be quite expensive.

**Comment by lundalana** [comment]
No worries! The explanation of the trajectories in your most recent comment (and in the paper) makes a lot of sense. 

With respect to (1) you are correct. For my case in particular, which is an application in civil engineering for structural health monitoring, having easy access to the marginal distributions on both the states and parameters is particularly helpful. I can also see this feature being useful for problems in control. So in answer to your question, I would say yes, it would be very useful for these marginals to be gathered in smc2.summaries. From a coding standpoint, I also think it would be a little more intuitive to have a summary of the progression of these marginals over time, as the method is designed to infer both the states and parameters. 

Thank you again for your help!
